### PR TITLE
Copy documentation from gh-pages to doc/contribute

### PR DESCRIPTION
**Pull request by sarakonradi**
This PR copies content from the ``gh-pages`` branch into ``doc/contribute``. It addresses the checklist items in #1659 related to the contributing guide.

More specifically, it
- moves the ``NEST_Contributor_Agreement.pdf`` to ``doc/contribute``
- converts the Git workflow and C++ coding guidelines to RST
- updates ``doc/contribute/index.rst``
- adds the "two reviewers" guideline to the contributing page

I did not remove too much from the Git workflow (yet), due to its references to NEST throughout. Please feel free to suggest which sections we can remove with clear conscience!

### CODE REVIEW

File: doc/contribute/coding_guidelines_cpp.rst [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+
+Use ``lower_case_under_lined`` notation for folder names.
+
+Variables and class members
+***************************
+
+In general, use meaningful, non-abbreviated names or follow naming conventions
+from the neuroscience field, e.g. the membrane potential is ``V_m``. Use the
+``lower_case_under_lined`` notation. Private member variables should end with an
+underscore (``name_``).
```
Comment (jougs) [inline doc/contribute/coding_guidelines_cpp.rst:357]:
```suggestion
underscore (``name_``). If applicable, the general rule is use is to use the
same notation for biophysical quantities as is used in `Dayan&Abbot, 2001
<https://www.gatsby.ucl.ac.uk/~lmate/biblio/dayanabbott.pdf>`_.
```

File: doc/contribute/coding_guidelines_cpp.rst [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+     G_EXC,
+     DG_INH,
+     G_INH,
+     STATE_VEC_SIZE
+   };
+
+Built-in types
+**************
+
+All code for the nest kernel should use the type aliases, defined in ``nest.h``.
```
Comment (jougs) [inline doc/contribute/coding_guidelines_cpp.rst:377]:
```suggestion
All code for the NEST kernel should use the type aliases, defined in ``nest.h``.
```

File: doc/contribute/coding_guidelines_cpp.rst [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+     timestamp_t elapsed_timestamp() const;
+
+     void reset();
+
+     void print( const char* msg = "",
+                 timeunit_t timeunit = SECONDS,
+                 std::ostream& os = std::cout ) const;
+
+     friend std::ostream& operator<<( std::ostream& os,
+                                      const Stopwatch& stopwatch );
```
Comment (jougs) [inline doc/contribute/coding_guidelines_cpp.rst:605]: Please check if these are below 120 characters if the line-breaks are removed. To be sure, it is maybe best to copy the current version of the `Stopwatch` files from `libnestutil` to here.

File: doc/contribute/development_workflow.rst [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+=================
+
+.. contents:: On this page, you'll find
+   :local:
+   :depth: 2
+
+Basic Git setup
+---------------
+
+The NEST development team uses git for version control.
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:11]:
```suggestion
The NEST development takes place in a repository using the Git version control system.
```

File: doc/contribute/development_workflow.rst [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+2. Go to the `NEST source code repository <https://github.com/nest/nest-simulator>`_ on GitHub.
+3. Click on the ``Fork`` button.
+
+After a short pause, you should find yourself at the home page for your own
+forked copy of NEST.
+
+Downloading your fork
+#####################
+
+After forking the repository, you need to download it to your local computer to
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:59]:
```suggestion
After forking the repository, you need to download (*clone*) it to your local computer to
```

File: doc/contribute/development_workflow.rst [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+This tells you that you are currently on the ``master`` branch, and that you
+also have a ``remote`` connection to ``origin/master``. The ``master`` branch is the
+default branch and this is where code that has been reviewed and tested resides. 
+``origin/master`` is just a copy of the ``master`` branch on your system on the ``remote``.
+
+What remote repository is ``remote/origin``? Try ``git remote -v`` to see the web
+address for the remote. It should point to your GitHub fork.
+
+Next, you connect your local copy to the central [NEST GitHub
+repository][NEST GitHub], so that you can keep your local copy and remote fork
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:97]: I've never seen this `[...][...]` syntax. What does it do? Is this a link to something? if yes, to what? Thanks!

File: doc/contribute/development_workflow.rst [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+up to date in the future. Conventionally, the main source code repository is
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:98]:
```suggestion
up to date in the future. By convention, the main source code repository is usually
```

File: doc/contribute/development_workflow.rst [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+
+   git status
+
+3. Review the changes you've made (optional).
+
+.. code::
+
+   git diff
+
+4. Inform git that you want to save these changes.
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:224]:
```suggestion
4. Inform Git that you want to save these changes.
```

File: doc/contribute/development_workflow.rst [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+   This brings up a simple text browser interface that highlights the difference
+   between your files and the previous version like this:
+
+   .. code::
+
+      diff --git a/development_workflow.md b/development_workflow.md
+      index f05f0cd..e581f00 100644
+      --- a/development_workflow.md
+     +++ b/development_workflow.md
+      @@ -8,17 +8,22 @@ layout: index
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:280]:
```suggestion
      diff --git a/development_workflow.rst b/development_workflow.rst
      index f05f0cd..e581f00 100644
      --- a/development_workflow.rst
     +++ b/development_workflow.rst
      @@ -8,17 +8,22 @@ layout: index
```

File: doc/contribute/development_workflow.rst [hunk 10]
```diff
@@ -0,0 +1,630 @@
...
+
+6. Inform Git of what modified or new files you want to save (stage) using ``git add modified_file``.
+   This puts the files into a ``staging area``, which is a
+   list of files that will be added to your next commit. Only add files that have
+   related, complete changes. Leave files with unfinished changes for later
+   commits.
+
+7. To commit the staged files into the local copy of your repo, do
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:288]:
```suggestion
7. To commit the staged files into the local copy of your repository, run 
```

File: doc/contribute/development_workflow.rst [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+
+8. Push the changes to your forked repo on GitHub:
+
+   .. code::
+
+      git push origin my-new-feature
+
+Assuming you have followed the instructions in these pages, git will create
+a default link to your GitHub repo called ``origin``. In git >= 1.7 you can
+ensure that the link to origin is permanently set by using the ``--set-upstream``
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:307]:
```suggestion
Assuming you have followed the instructions in these pages, Git will create
a default link to your GitHub repository called ``origin``. In Git >= 1.7 you can
ensure that the link to origin is permanently set by using the ``--set-upstream``
```

File: doc/contribute/development_workflow.rst [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+option:
+
+.. code::
+
+   git push --set-upstream origin my-new-feature
+
+From now on, Git will know that ``my-new-feature`` is related to the
+``my-new-feature`` branch in your own GitHub repo. Subsequent push calls
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:315]:
```suggestion
``my-new-feature`` branch in your own GitHub repository. Subsequent push calls
```

File: doc/contribute/development_workflow.rst [hunk 13]
```diff
@@ -0,0 +1,630 @@
...
+
+.. _pull_request:
+
+Creating a pull request
+~~~~~~~~~~~~~~~~~~~~~~~
+
+When you feel your work is finished, you can create a pull request (PR). GitHub
+has a nice help page that outlines the process for
+`submitting pull requests <https://help.github.com/articles/using-pull-requests/#initiating-the-pull-request>`_.
+Your pull request will usually be reviewed by other NEST developers using the code review guidelines.
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:337]: Please add a link to the code review guidelines and to the C++ style guide.

File: doc/contribute/development_workflow.rst [hunk 14]
```diff
@@ -0,0 +1,630 @@
...
+
+Advanced cases
```
Comment (jougs) [inline doc/contribute/development_workflow.rst:339]: I would like to vote for removing the advanced cases altogether. They make the document only longer without a benefit to most people and all core developers should be able to solve these kinds of problems on their own.

File: doc/contribute/eclipse_workflow.rst [hunk 15]
```diff
@@ -0,0 +1,346 @@
+Developing NEST with Eclipse
+============================
+
+These instructions are based on Eclipse Mars (4.5). They are based
+on earlier instructions by Thomas Heiberg.
```
Comment (jougs) [inline doc/contribute/eclipse_workflow.rst:5]:
```suggestion
These instructions are based on Eclipse Mars (4.5).
```

File: doc/contribute/eclipse_workflow.rst [hunk 16]
```diff
@@ -0,0 +1,346 @@
+Developing NEST with Eclipse
```
Comment (jougs) [inline doc/contribute/eclipse_workflow.rst:1]: I nominate @heplesser to review the Eclipse Development documentation, as he is actually using it.

File: doc/contribute/xcode_workflow.rst [hunk 17]
```diff
@@ -0,0 +1,130 @@
+Xcode workflow
```
Comment (jougs) [inline doc/contribute/xcode_workflow.rst:1]: @heplesser?

File: doc/contribute/development_workflow.rst (hunk 14 above)
Comment (sarakonradi) [reply doc/contribute/development_workflow.rst]: Yes, I agree and have therefore removed the advanced cases. If anything, the information is also available in the Git documentation.

File: doc/contribute/development_workflow.rst (hunk 6 above)
Comment (sarakonradi) [reply doc/contribute/development_workflow.rst]: Ah, well spotted! This is a reference-style link in Markdown. It doesn't do anything in RST ;) I have now fixed it.

**Comment by sarakonradi** [comment]
@jougs Thanks for the detailed review! I have applied your suggestions and replied to your comments inline.

**Comment by heplesser** [comment]
@sarakonradi It is going to take me some time to review (and revise) the Eclipse workflow. I'd suggest to drop it from this PR and create a separate issue for adding it. I am about to recreate my Eclipse setup, so I will definitely update the instructions.

Concerning XCode, maybe @tammoippen could chip in?

**Comment by jougs** [comment]
Just for the record: I'd be totally fine with removing the two IDE related pages for now.

**Comment by sarakonradi** [comment]
>  I'd suggest to drop it from this PR and create a separate issue for adding it. I am about to recreate my Eclipse setup, so I will definitely update the instructions.

@heplesser Sounds good. I have created a separate issue for the Eclipse workflow and undone the review request here.

**Comment by jougs** [comment]
@sarakonradi: many thanks!

As this is "only" concerning documentation and really just a move of existing pages, I think we can do with a single review here. I'm thus merging.

**Pull request merged**